### DISCUSSION TITLE
Using relay controls

### DISCUSSION THREAD

**Discussion by axg20202** [phil-barrett/grblHAL-teensy-4.x#73] [General]
Hi Phil. I can't seem to get my external relays to trigger. I am powering the board by the 5v screw terminals, no USB connected (I'm using ethernet). Do I also need to inject 5v at the 12v/5v screw terminals at the other side of the board to get the control outputs firing? I currently have nothing connected at that power input. Jumper is set to 5v. Vac/DC digital output signal pin is connected to the signal input of my relay (has its own coil power). It's not working. Any ideas?

**Comment by phil-barrett** [comment]
If you are directly driving relay coils, have you jumpered the relay voltage? It's in the upper left hand corner of the board. 

![image](https://user-images.githubusercontent.com/16570958/213873971-cfefd846-4001-4300-a1cc-fe53fe8e7829.png)

**Reply by axg20202** [reply]
Hi Phil. I have the jumper at 5v position, but no power attached to the terminals. I'm not driving coils directly, just triggering the signal input on the relay board, which has its own coil power.

**Comment by phil-barrett** [comment]
OK, So you are using the TTL/SSR outputs.  If the LED in the lower left corner lights up, you have 5V power. Got a ground connection between the relay board and the T41U5XBB?  Could you hook up a DVM to the T41U5XBB pins to see if there is voltage when you activate the appropriate M command?  

Also, have you verified that the relay board works?  Try running a gnd and +5 wire from the T41U5XBB to the relay inputs.

**Reply by axg20202** [reply]
Thanks Phil. I will try running a ground between the relay board and BOB (Vac TTL gnd pin). However, the relay board and the BOB are being powered by the same 5V DC supply, so I had assumed their grounds would already be connected?

**Comment by axg20202** [comment]
OK, I've worked out what's happening, but not solved the issue. The Vac TTL output isnt outputting a 5v signal when I start the spindle. The other TTL pins work (Mist, Flood) and switch the relay no problem, simply by ticking them on in IOsender. So, I'm guessing it's something to do with how my spindle is started/enabled?  My spindle is being controlled by the serial outputs of the BOB, not by the spindle connections. Not sure why this would make a difference as I'm still sending an M3 gcode to start the spindle.

**Reply by phil-barrett** [reply]
Hmmm.  Check the spindle enable (En) signal at the screw terminals. That should be the same as Vac.  Also, try shorting the Vac/DC Trigger header and see what the Vac output does.  

Are you using modbus to control the spindle? 

A longshot - check the soldering of the TTL/SSR header. Look for shorts as well as missing solder. Since the other relay outputs work, a bad chip is unlikely but a shorted pin could blow a single output on the driver chip.

Also, what are you using for firmware - your own build or one of the prebuilt ones?

**Reply by axg20202** [reply]
Thanks for your help, Phil.  I just tested the spindle enable pins and they also don't output anything when the spindle is turned on. Yes, I'm using modbus. Shorting the Vac trigger header does cause 5v to appear at the Vac TTL pins. Leads me to suspect that it might be a firmware thing where spindle enable isn't being activated, but not sure.

I compiled the firmware myself. All else is working as it should.

**Reply by phil-barrett** [reply]
OK, the fact that the vac/dc header works means the hardware is fine.  I took a look at the Spindle/Modbus plugin but didn't see anything about the spindle enable output.

You might try loading FW that doesn't have the spindle plugin to see if M3 causes the spindle enable to go high. 

You might want to ask about the plugin at the [grblHAL core github issues page](https://github.com/grblHAL/core/issues).

**Reply by axg20202** [reply]
Thanks for your help, Phil - much appreciated. I asked there and apparently this is expected behaviour when using modbus. Hopefully I can find a way to change this and recompile.

**Reply by phil-barrett** [reply]
No problem.  Wish you got a better answer.

**Comment by axg20202** [comment]
Hi Phil. Would it be OK to connect two TTL signal pins so that they always go high together? Because of the Vac issue, I plan to use the Aux as the automatic trigger for Vacuum, but I'd like to retain the manual Vac trigger. If I link their TTL signal pins, this should send 5v to the Vac TTL output when triggered by either Aux, or using the manual trigger. Just wanted to check that this won't cause an issue or damage. Would rather not have to use multiple parallel relays to achieve this.

**Reply by phil-barrett** [reply]
Sorry for not getting back sooner.   

No, I would not connect two TTL (totem pole/push pull) outputs together.  That would create short circuit when one is high and the other low.  Now, if you use the relay outputs on the T41U5XBB, those are open collector so you could do it there.  If you are driving a logic input (like on a "TTL" input relay board), you will need pull up resistors. If you are directly driving relay coils, you can just connect both outputs to the relay coil

By the way, Terje has made a fix to the fans module that should let you get spindle enable to go high. I haven't dug into yet but sounds promising.
